daemon.go

```diff
@@ -2,170 +2,94 @@ package main
 
 import (
 	"context"
-	"encoding/json"
 	"os"
 	"os/exec"
-	"time"
+	"sync"
 
-	"github.com/aws/aws-sdk-go/aws"
-	"github.com/aws/aws-sdk-go/service/autoscaling"
-	"github.com/aws/aws-sdk-go/service/sqs"
 	log "github.com/sirupsen/logrus"
 )
 
-const (
-	heartbeatFrequency = time.Second * 10
-)
-
-type Envelope struct {
-	Type    string    `json:"Type"`
-	Subject string    `json:"Subject"`
-	Time    time.Time `json:"Time"`
-	Message string    `json:"Message"`
-}
+type Daemon struct {
+	LifecycleMonitor *LifecycleMonitor
+	SpotMonitor      *SpotMonitor
 
-type AutoscalingMessage struct {
-	Time        time.Time `json:"Time"`
-	GroupName   string    `json:"AutoScalingGroupName"`
-	InstanceID  string    `json:"EC2InstanceId"`
-	ActionToken string    `json:"LifecycleActionToken"`
-	Transition  string    `json:"LifecycleTransition"`
-	HookName    string    `json:"LifecycleHookName"`
+	InstanceID string
+	Handler    *os.File
 }
 
-type Daemon struct {
-	InstanceID  string
-	Queue       *Queue
-	AutoScaling *autoscaling.AutoScaling
-	Handler     *os.File
+type TerminationNotice struct {
+	Done  chan struct{}
+	Error chan error
+	Args  []string
 }
 
 // Start the daemon.
 func (d *Daemon) Start(ctx context.Context) error {
-	if err := d.Queue.Create(); err != nil {
-		return err
-	}
-	defer func() {
-		if err := d.Queue.Delete(); err != nil {
-			log.WithError(err).Error("Failed to delete queue")
-		}
-	}()
-
-	if err := d.Queue.Subscribe(); err != nil {
-		return err
-	}
-	defer func() {
-		if err := d.Queue.Unsubscribe(); err != nil {
-			log.WithError(err).Error("Failed to unsubscribe from sns topic")
-		}
-	}()
-
-	ch := make(chan *sqs.Message)
+	var wg sync.WaitGroup
 
-	go func() {
-		for m := range ch {
-			var env Envelope
-			var msg AutoscalingMessage
-
-			// unmarshal outer layer
-			if err := json.Unmarshal([]byte(*m.Body), &env); err != nil {
-				log.WithError(err).Info("Failed to unmarshal envelope")
-				continue
-			}
+	var errCh = make(chan error, 2)
+	var termCh = make(chan TerminationNotice)
 
-			log.WithFields(log.Fields{
-				"type":    env.Type,
-				"subject": env.Subject,
-			}).Debugf("Received an SQS message")
+	subctx, cancel := context.WithCancel(ctx)
+	defer cancel()
 
-			// unmarshal inner layer
-			if err := json.Unmarshal([]byte(env.Message), &msg); err != nil {
-				log.WithError(err).Info("Failed to unmarshal autoscaling message")
-				continue
-			}
+	// Process lifecycle events
+	if d.LifecycleMonitor != nil {
+		wg.Add(1)
 
-			if msg.InstanceID != d.InstanceID {
-				log.WithFields(log.Fields{
-					"was":    msg.InstanceID,
-					"wanted": d.InstanceID,
-				}).Debugf("Skipping autoscaling event, doesn't match instance id")
-				continue
+		go func() {
+			defer wg.Done()
+			if err := d.LifecycleMonitor.Run(subctx, termCh); err != nil {
+				log.WithError(err).Debug("Lifecycle monitor failed with error")
+				errCh <- err
```

Comment:
In my opinion, we don't need to pass the error over `errCh` because it was already handled by being logged and cancelling the context?